File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> map;
    
    void dfs(TreeNode* root, int level) {
        if(!root)   return;
        
        map[level] += root->val;
        
        dfs(root->left, 1 + level);
        dfs(root->right, 1 + level);
    }
    
    int maxLevelSum(TreeNode* root) {
        dfs(root, 1);
        int sum = INT_MIN;
        int ans = 0;
        
        for(auto i = map.begin(); i != map.end(); i++) {
            if(i->second > sum) {
                sum = i->second;
                ans = i->first;
            } else if(i->second == sum) {
                ans = min(ans, i->first);
            }
        }
        
        return ans;
    }
};

File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxLevelSum(TreeNode root) {
        HashMap<Integer, Integer> map = new HashMap<>();
        dfs(root, map, 1);
        int maxVal = Integer.MIN_VALUE;
        int ans = 0;

        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if(entry.getValue() > maxVal) {
                maxVal = entry.getValue();
                ans = entry.getKey();    
            }
        }

        return ans;
    }

    private void dfs(TreeNode root, HashMap<Integer, Integer> map, int level) {
        if (root == null)
            return;
        map.put(level, map.getOrDefault(level, 0) + root.val);
        dfs(root.left, map, level + 1);
        dfs(root.right, map, level + 1);
    }
}